### PR TITLE
Create integration tests with espresso framework  

### LINKED ISSUE #14: Create integration tests
- [Testing fundamentals](https://stuff.mit.edu/afs/sipb/project/android/docs/tools/testing/testing_android.html)
- [Tests The Android Way](https://www.bignerdranch.com/blog/testing-the-android-way/)
- [Espresso Test Life Cycle](https://jabknowsnothing.wordpress.com/2015/11/05/activitytestrule-espressos-test-lifecycle/)
- [Tests with Espresso, Mockito, Dagger 2](https://engineering.circle.com/instrumentation-testing-with-dagger-mockito-and-espresso-f07b5f62a85b#.5rti2kl2e)
- [Espresso Tutorial](http://www.vogella.com/tutorials/AndroidTestingEspresso/article.html#espresso_exercisesimple)
- [Roboeletric vs Android Test Framework](http://stackoverflow.com/questions/18271474/robolectric-vs-android-test-framework)

### PR DESCRIPTION
**Pull request by niltonvasques**
- Test the gif play/pause logic
- Test loading show hide logic
- Test watched/not watched logic
- Test gif load logic
- Decoupling components to allow mock
- Test download failure using mocks

Fixes #14

### CODE REVIEW

**Comment by niltonvasques** [comment]
In CI Server it is possible to use this command to run all test suite:

```
  adb shell am instrument -w br.com.catbag.giffluxsample.test/android.support.test.runner.AndroidJUnitRunner
```

File: README.md [hunk 1]
```diff
@@ -17,6 +17,17 @@ A Android sample that implements Flux architecture with React and NoSQL
...
 ```
 
+#Tests
+
+###Integration tests
+
+- The integration tests is running on top of espresso framework, and using 
+mockito to handle mocks.
+
+To setup a CI or run tests in command line use this command:
```
Comment (raulccabreu) [inline README.md:34]: To run without CI, locally...i have to call **connectCheck** gradle task? or **check**? or simply build and run?

File: app/build.gradle [hunk 2]
```diff
@@ -30,6 +30,10 @@ dependencies {
     androidTestCompile('com.android.support.test.espresso:espresso-core:2.2.2', {
         exclude group: 'com.android.support', module: 'support-annotations'
     })
+    androidTestCompile 'com.google.dexmaker:dexmaker-mockito:1.0'
+    androidTestCompile 'com.google.dexmaker:dexmaker:1.0'
+    androidTestCompile "org.mockito:mockito-core:1.+"
```
Comment (raulccabreu) [inline app/build.gradle]: It's dangerous compile a dependency with "+" (all subversions)...

File: app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        };
+    }
+
+    public static Matcher<View> withBGColor(final int color) {
+        Checks.checkNotNull(color);
+        return new BoundedMatcher<View, View>(View.class) {
+            @Override
+            public boolean matchesSafely(View warning) {
+                int currentColor = ((ColorDrawable)warning.getBackground()).getColor();
+                System.out.println("Current color: "+currentColor);
```
Comment (raulccabreu) [inline app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java]: This system.out is needed yet?

File: app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                return color == currentColor;
+            }
+            @Override
+            public void describeTo(Description description) {
+                description.appendText("with background color: "+color);
```
Comment (raulccabreu) [inline app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java]: It's good code styte pattern put spaces between "+" operator.

File: app/src/androidTest/java/br/com/catbag/giffluxsample/ui/giflist/GifListActivityTest.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                .perform(click());
+        assert(!mActivityTestRule.getActivity().getGlideWrapper().getResource().isRunning());
+    }
+
+    private FileDownloader.FailureDownloadListener listener;
+    private FileDownloader downloader;
+    public void mockFileDownloaderToAlwaysFail(){
+        downloader = mock(FileDownloader.class);
+        doAnswer( (invocation) -> {
+            System.out.println("MOCKED METHOD");
```
Comment (raulccabreu) [inline app/src/androidTest/java/br/com/catbag/giffluxsample/ui/giflist/GifListActivityTest.java]: needed?

File: app/src/androidTest/java/br/com/catbag/giffluxsample/ui/giflist/GifListActivityTest.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            listener.onFailure(new Exception("Download error"));
+            return null;
+        }).when(downloader).download(anyString(), anyString());
+        doAnswer( (invocation) -> {
+            Object[] args = invocation.getArguments();
+            Object mock = invocation.getMock();
+            System.out.println("MOCKED FAILURE");
```
Comment (raulccabreu) [inline app/src/androidTest/java/br/com/catbag/giffluxsample/ui/giflist/GifListActivityTest.java]: Sysout needed?

File: app/src/main/java/br/com/catbag/giffluxsample/asyncs/restservice/FileDownloader.java [hunk 7]
```diff
@@ -18,9 +18,10 @@
 public class FileDownloader {
     private static final String TAG = "FileDownloader";
     private final ApiRoutes mRoutes;
-    private SuccessDownloadListener mSuccessListener = null;
-    private FailureDownloadListener mFailureListener = null;
-    private StartDownloadListener mStartListener = null;
+    private SuccessDownloadListener mSuccessListener;
+
```
Comment (raulccabreu) [inline app/src/main/java/br/com/catbag/giffluxsample/asyncs/restservice/FileDownloader.java]: Line needed?

File: app/src/main/java/br/com/catbag/giffluxsample/ui/GifListActivity.java [hunk 8]
```diff
@@ -94,6 +97,11 @@ public void onStateChanged(AppState appState) {
         Anvil.render();
     }
 
+    public GlideWrapper getGlideWrapper() {
+        return mGlideWrapper;
+    }
+
+
```
Comment (raulccabreu) [inline app/src/main/java/br/com/catbag/giffluxsample/ui/GifListActivity.java]: Line needed?

File: app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeMatcher;
+
+/**
+ * Created by niltonvasques on 10/14/16.
+ */
+
+public class Matchers {
+    private static Matcher<View> childAtPosition(
```
Comment (raulccabreu) [inline app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java]: This matcher it's not used...but will be used some time?

File: app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                return color == currentColor;
+            }
+            @Override
+            public void describeTo(Description description) {
+                description.appendText("with background color: "+color);
+            }
+        };
+    }
+
+    public static Matcher<View> animationRunning() {
```
Comment (raulccabreu) [inline app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java:35]: This matcher it's not used...but will be used some time?

**Review by raulccabreu — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (niltonvasques) [reply README.md:34]: Simply build and run!

File: app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java (hunk 10 above)
Comment (niltonvasques) [reply app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java:35]: I've wrote some matchers to help test the UI. This I wrote to test gif animation, but I've discovered that Glide doesn't uses animations to play gif. On this project, if we have some animations we will eventually need use this matcher.

File: app/src/androidTest/java/br/com/catbag/giffluxsample/ui/giflist/GifListActivityTest.java (hunk 5 above)
Comment (niltonvasques) [reply app/src/androidTest/java/br/com/catbag/giffluxsample/ui/giflist/GifListActivityTest.java]: No.

File: app/src/androidTest/java/br/com/catbag/giffluxsample/ui/giflist/GifListActivityTest.java (hunk 6 above)
Comment (niltonvasques) [reply app/src/androidTest/java/br/com/catbag/giffluxsample/ui/giflist/GifListActivityTest.java]: No.

File: app/src/main/java/br/com/catbag/giffluxsample/asyncs/restservice/FileDownloader.java (hunk 7 above)
Comment (niltonvasques) [reply app/src/main/java/br/com/catbag/giffluxsample/asyncs/restservice/FileDownloader.java]: No.

File: app/src/main/java/br/com/catbag/giffluxsample/ui/GifListActivity.java (hunk 8 above)
Comment (niltonvasques) [reply app/src/main/java/br/com/catbag/giffluxsample/ui/GifListActivity.java]: No.

File: app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java (hunk 4 above)
Comment (niltonvasques) [reply app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java]: I will fix it.

File: app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java (hunk 3 above)
Comment (niltonvasques) [reply app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java]: No

File: app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java (hunk 9 above)
Comment (niltonvasques) [reply app/src/androidTest/java/br/com/catbag/giffluxsample/matchers/Matchers.java]: Probably not.

File: app/build.gradle (hunk 2 above)
Comment (niltonvasques) [reply app/build.gradle]: You are right.

**Review by raulccabreu — APPROVED** [review]

**Pull request merged**